fireplace/dsl/random_picker.py

```diff
@@ -1,79 +1,96 @@
 import random
 from hearthstone.enums import CardType, Race, Rarity
 from .lazynum import LazyValue
+from copy import deepcopy
+from .. import cards
 
 
 class RandomCardPicker(LazyValue):
 	"""
 	Store filters and generate a random card matching the filters on pick()
+	Constructor takes a single global set of filters and uses a default weighting of 1
+	Additional weighted filter sets can be added with add(), these will be merged with the global filters
 	"""
-	def __init__(self, *args, **filters):
-		self.args = args
-		self.filters = filters
+	def __init__(self, **filters):
+		self.weights = []
+		self.weightedfilters = []
+		self.globalfilter = filters
 		self.count = 1
-		self._cards = None
-		self.lazy_filters = False
-		for v in filters.values():
-			if isinstance(v, LazyValue):
-				self.lazy_filters = True
-				break
 
 	def __repr__(self):
-		return "%s(%r)" % (self.__class__.__name__, self.filters)
+		return "%s(%r)" % (self.__class__.__name__, self.globalfilter)
 
+	# select number of cards to fetch
 	def __mul__(self, other):
-		ret = self.__class__(*self.args, **self.filters)
+		ret = deepcopy(self)
 		ret.count = other
 		return ret
 
-	@property
-	def cards(self):
-		if self._cards is None:
-			self._cards = self._filter_cards(self.filters)
-		return self._cards
+	# add a filter set
+	def add(self, weight, **filters):
+		self.weights.append(weight)
+		self.weightedfilters.append(filters)
+		return self
 
-	def _filter_cards(self, filters):
-		from .. import cards
-		return cards.filter(**filters)
+	# generate a card pool with all cards matching specified (optionally lazy) filters
+	def find_cards(self, source=None, **filters):
+		if len(filters)==0:
+			newFilters = self.globalfilter.copy()
+		else:
+			newFilters = filters.copy()
 
-	def get_cards(self, source):
-		filters = self.filters.copy()
-		# Iterate through the filters, evaluating the LazyValues as we go
-		for k, v in filters.items():
+		for k, v in newFilters.items():
 			if isinstance(v, LazyValue):
-				filters[k] = v.evaluate(source)
-		return self._filter_cards(filters)
+				newFilters[k] = v.evaluate(source)
+
+		return cards.filter(**newFilters)
+
+	# This picks from a single combined card pool without replacement, weighting each filtered set of cards against the total
+	def evaluate(self, source, cards=None) -> str:
+
+		# use specific card list if given
+		if cards:
+			self.weights = [1]
+			card_sets = [cards]
 
-	def evaluate(self, source) -> str:
-		if self.lazy_filters:
-			# If the card has lazy filters, we need to evaluate them
-			cards = self.get_cards(source)
+		# use global filters if no weighted filter sets given
```

Comment:
Is there a strong reason to distinguish between a global filter and filter sets? If not, I'd advocate just using a singleton filter set with any weight.